mirgecom/filter.py

```diff
@@ -0,0 +1,194 @@
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+"""
+
+import numpy as np
+import loopy as lp
+from grudge import sym
+from meshmode.dof_array import DOFArray
+from pytools import memoize_in
+from pytools.obj_array import (
+    obj_array_vectorized_n_args,
+    #    obj_array_vectorize
+)
+
+
+def exponential_mode_response_function(mode, alpha, cutoff, nfilt, filter_order):
+    """Return an exponential filter coefficient for the user-provided *mode*."""
+    return np.exp(-1.0 * alpha * ((mode - cutoff) / nfilt)
+                  ** (2*filter_order))
+
+
+def make_spectral_filter(element_group, cutoff, mode_response_function):
+    r"""
+    Create a spectral filter with the provided *mode_response_function*.
+
+    This routine returns a filter operator in the modal basis designed to
+    apply the user-provided *mode_response_function* to the spectral modes
+    beyond the user-provided *cutoff*.
+
+    Parameters
+    ----------
+    element_group: :class:`meshmode.mesh.MeshElementGroup`
+        A :class:`meshmode.mesh.MeshElementGroup` from which the mode ids,
+        element order, and dimension may be retrieved.
+    cutoff: integer
+        Mode cutoff beyond which the filter will be applied, and below which
+        the filter will preserve.
+    mode_response_function:
+        A function that returns a filter weight for for each mode id.
+
+    Returns
+    -------
+    filter: np.ndarray
+        filter operator in the modal basis
+    """
+    mode_ids = element_group.mode_ids()
+    order = element_group.order
+    dim = element_group.dim
+
+    nmodes = len(mode_ids)
+    filter = np.identity(nmodes)
+    nfilt = order - cutoff
+    if nfilt <= 0:
+        return filter
+
+    for mode_index, mode_id in enumerate(mode_ids):
+        mode = mode_id
+        if dim > 1:
+            mode = sum(mode_id)
+        if mode >= cutoff:
+            filter[mode_index, mode_index] = mode_response_function(mode,
+                                                                    cutoff=cutoff,
+                                                                    nfilt=nfilt)
+    return filter
+
+
+def linear_operator_kernel():
+    """Apply linear operator to all elements."""
+    from meshmode.array_context import make_loopy_program
+    knl = make_loopy_program(
+        """{[iel,idof,j]:
+        0<=iel<nelements and
+        0<=idof<ndiscr_nodes_out and
+        0<=j<ndiscr_nodes_in}""",
+        "result[iel,idof] = sum(j, mat[idof, j] * vec[iel, j])",
+        name="spectral_filter")
+    knl = lp.tag_array_axes(knl, "mat", "stride:auto,stride:auto")
+    return knl
+
+
+@obj_array_vectorized_n_args
+def apply_linear_operator(discr, operator, fields):
```

Comment:
A (meshmode) discretization can consist of multiple element groups, which may have different element types. The same matrix is unlikely to cut it for all of them, since they're not even guaranteed to have the same DOF count.

Since it's conceptually wrong, I kind of don't want it in the public API. I see that it's only used in the tests. I'd be happier with it if it were moved to the tests and asserted that `discr` only has one group.